File: python3/minimum_falling_path_sum.py
"""
Space   : O(1)
Time    : O(nm)
"""


class Solution:
    def minFallingPathSum(self, A: List[List[int]]) -> int:
        row = len(A)
        col = len(A[0])

        if row == 1:
            return min(A[0])
        if col == 1:
            ans = 0
            for i in range(row):
                ans += A[i][0]
            return ans

        for y in range(1, row):
            for x in range(col):
                if x == 0:
                    A[y][x] += min(A[y-1][x], A[y-1][x+1])
                elif x == col-1:
                    A[y][x] += min(A[y-1][x], A[y-1][x-1])
                else:
                    A[y][x] += min(A[y-1][x-1], A[y-1][x], A[y-1][x+1])

        return min(A[-1])

File: golang/minimum_falling_path_sum.go
package main

/*
Space	: O(1)
Time	: O(nm)
*/

func minFallingPathSum(matrix [][]int) int {
	var ans int
	n := len(matrix)

	if n == 1 {
		return matrix[0][0]
	}

	for i := 1; i < n; i++ {
		for j := 0; j < n; j++ {
			var minPrev int

			if j == 0 {
				minPrev = Min(matrix[i-1][j], matrix[i-1][j+1])
			} else if j == n-1 {
				minPrev = Min(matrix[i-1][j-1], matrix[i-1][j])
			} else {
				minPrev = Min(matrix[i-1][j-1], matrix[i-1][j])
				minPrev = Min(minPrev, matrix[i-1][j+1])
			}

			matrix[i][j] += minPrev
		}
	}

	for i := n - 1; i < n; i++ {
		for j := 0; j < n; j++ {
			if j == 0 {
				ans = matrix[i][j]
			} else {
				ans = Min(ans, matrix[i][j])
			}
		}
	}

	return ans
}
